### ISSUE TITLE
SurfacePlotVisual.set_data cause vertex shader  to perform unnecessary build

### ISSUE THREAD

**Issue by Akalivivy** [vispy/vispy#2568]
when I try to  use scene.visuals.SurfacePlot.set_data with an app.Timer to get some animation effects,then I found that the CPU usage was very high.My core code like :

``` 
     # some init code in my class
     empty_data = np.zeros((13, 14))
     z = gaussian_filter(empty_data, (10, 10))
     self.plot = scene.visuals.SurfacePlot(z=empty_data)
     no_colors = np.ones((13, 14, 3))
     self.plot.set_data(z=empty_data, colors=no_colors)
     self.view.add(self.plot)
    # timer
     self.timer = app.Timer('auto', self.on_timer)
     self.timer.start(1/60)
    
    # the core update function
    def on_timer(self,event):
        print("on timer begin update")
        colors = np.zeros((13, 14, 3))
        shape = (13,14)  
        visualize_data = 3*np.random.rand(*shape)
        self.plot.set_data(z=visualize_data, colors=colors)
        print("on timer update have completed")


```
Then I traced the call process of the function from the source code，I found when the Superclass of SurfacePlotVisual,MeshVisual,calls the following code cause the vertex shader rebuild:
`self.shared_program.vert['color_transform'] = self._build_color_transform(colors)`
the line of mesh.py source code here:[mesh.py](https://github.com/vispy/vispy/blob/main/vispy/visuals/mesh.py#L346)

`fun = Function(null_color_transform)` ([build_color_transform](https://github.com/vispy/vispy/blob/main/vispy/visuals/mesh.py#L290)) in the the fuction `_build_color_transform` create new Function object,when it call `__setitem__` in the class Function [function.py](https://github.com/vispy/vispy/blob/main/vispy/visuals/shaders/function.py#L190),the self.changed check process  think  both code_changed and value_changed,In the subsequent process, shared_program (shader) will be affected by the `code_changed` and believed that it need build([program.py](https://github.com/vispy/vispy/blob/main/vispy/visuals/shaders/program.py#L87)),though the code for vertex shaders has not actually changed(the default text of null_color_transform:[null_color_transform](https://github.com/vispy/vispy/blob/main/vispy/visuals/mesh.py#L282) ).

So,it seems like once I use the SurfacePlotVisual.set_data, it cause  vertex shader build.But I just hope to use `set_data` to replace the visual data during the loop process, but I cannot avoid rebuilding the shader.Could someone help me or had a similar experience？

**Comment by djhoese** [comment]
I agree. I think you're correct. I think this visual has not been optimized for this color logic which is surprising. I think @brisvag or maybe @tlambert03 did some work on the ImageVisual (https://github.com/vispy/vispy/blob/b72a201ed7e00c34d2ec676d23d45e3c62f618ed/vispy/visuals/image.py#L678-L682) to optimize when and how the color transform function is created and updated. I think the MeshVisual may need a similar update and extra care added to it.

I think the original intention of the code or the idea behind it was that if you change the data then the colors array will also change (I'm not sure this is always necessary either). Since `colors` is an input to the color transform creation it would make sense to assume that the color transform also always has to be updated. But you're right, when you look at it it clearly doesn't.

First, it would need to be updated so the function objects only get created once. I don't think there is a reason this shouldn't work. Next, it should be updated so if we already have a color transform but we are updating only the color limits, then just update the color limits. I think mocking this after the ImageVisual wouldn't be a terrible idea. This would mean something like a `self._update_color_transform = True` and maybe a `self._update_color_limits = True`. You can see in the `ImageVisual` that it even cheats and sets the color limits early if the color transform doesn't need to be updated:

https://github.com/vispy/vispy/blob/b72a201ed7e00c34d2ec676d23d45e3c62f618ed/vispy/visuals/image.py#L419-L429

**Comment by brisvag** [comment]
Yup, agree with the above! I don't have the bandwidth now to do with this, but if anyone wants to open a PR with this I'd be happy to review.

**Comment by Akalivivy** [comment]
@djhoese Thank you for your answer. In meshvisual, it is indeed necessary to use optimization logic similar to that already present in ImageVisual to avoid the repeated creation of function object。If I have time, I would be happy to try and contribute to the code.
